tensorflow_similarity/losses/lifted_structure_loss.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Lifted Structured Loss
+    Deep Metric Learning via Lifted Structured Feature Embedding.
+    https://arxiv.org/abs/1511.06452
+"""
+
+import tensorflow as tf
+
+from tensorflow_similarity.algebra import build_masks
+from tensorflow_similarity.distances import Distance, distance_canonicalizer
+from tensorflow_similarity.types import FloatTensor, IntTensor
+from tensorflow_similarity import losses as tfsim_losses
+from .metric_loss import MetricLoss
+from .utils import negative_distances, positive_distances
+
+
+def lifted_struct_loss(
+    labels: IntTensor,
+    embeddings: FloatTensor,
+    distance: Distance,
+    positive_mining_strategy: str = "hard",
+    negative_mining_strategy: str = "easy",
+    margin: float = 1.0,
+) -> FloatTensor:
+    """Lifted Struct loss computations"""
+
+    # Compute pairwise distances
+    pairwise_distances = distance(embeddings)
+
+    # Build masks for positive and negative pairs
+    positive_mask, negative_mask = build_masks(query_labels=labels, batch_size=tf.shape(embeddings)[0])
+
+    # Get positive distances and indices
+    positive_dists, positive_indices = positive_distances(
+        positive_mining_strategy, pairwise_distances, positive_mask
+    )
+
+    # Get negative distances
```

Comment:
Looking at this again, I think we can remove lines 52-54 and use `pairwise_distances` in the tf.gather on line 57.

My concern is that the negative mining strategy is meant to return a single negative per row, but here we really want to keep all the negatives.

wdyt?

Reply:
Yeah I mean, it also makes more sense to do this in `tf.gather` so we keep consistency (so something like `reordered_pairwise_distances = tf.gather(pairwise_distances, positive_indices, axis=1)`) which also agrees with `reordered_pairwise_distances`. As fair as I understand by reading the paper, I do think we want to keep all the negatives.

Do you think it makes sense to remove neg mining strategy at al?